### DISCUSSION TITLE
Data retention for assets (partitioned and not-partitioned)

### DISCUSSION THREAD

**Discussion by cleboo** [dagster-io/dagster#15381] [Q&A] [answered]
Hi!

I have a couple of questions regarding data retention of assets in dagster:
+ is there a way to define how long an asset and its data should be kept? (e.g.: delete the asset and the associated data after 1 month)
+ is it possible to change the start date of a partition? What happens to the already existing assets if a partition_key is removed?

The use-case I am thinking of is to have a daily partition that runs over the last 30 days and that the assets and the associated files are  removed when an asset ages out.

In general, how does everybody manage deleting old assets and data?

**Comment by clairelin135** [comment, marked as answer]
Hi @cleboo. These are good questions.

> Is there a way to define how long an asset and its data should be kept? (e.g.: delete the asset and the associated data after 1 month)

We do not have built-in ways to do this, though I'd recommend filing an issue if this is something you're interested in. One idea here is to add a CLI command that allows you to wipe certain partitions of an asset (related issue https://github.com/dagster-io/dagster/issues/14749). 

> is it possible to change the start date of a partition? What happens to the already existing assets if a partition_key is removed?

If you change the start date of your partitions definition, you will just no longer see the history for the partitions that are no longer part of your definition. The history doesn't get deleted though, so if you changed the start date back, you would see it again. 

In your use case, you can change the start date of your partitions def to just be thirty days ago, so it only represents the data from the last thirty days. You could set up a schedule that runs daily that deletes any associated files older than thirty days.

Related: if you did want to delete the history of your assets (i.e. run logs, data about which partitions were materialized, etc.) this discussion details how you can do this: https://github.com/dagster-io/dagster/discussions/12047

**Reply by cleboo** [reply]
Thanks @clairelin135!

> If you change the start date of your partitions definition, you will just no longer see the history for the partitions that are no longer part of your definition. The history doesn't get deleted though, so if you changed the start date back, you would see it again.
> In your use case, you can change the start date of your partitions def to just be thirty days ago, so it only represents the data from the last thirty days. You could set up a schedule that runs daily that deletes any associated files older than thirty days.

Yes, I tried that out yesterday by using a dynamic partition and truncating the dates in a scheduled job. Do you know a reliable way to find the files associated with assets that have no partition anymore? I guess I'd have to dig through the history of the asset?

> One idea here is to add a CLI command that allows you to wipe certain partitions of an asset (related issue https://github.com/dagster-io/dagster/issues/14749).

That would be great!

> We do not have built-in ways to do this, though I'd recommend filing an issue if this is something you're interested in.

I'll put to together a feature request. My idea is that one can set a policy on an asset that determines how long the data is kept (forever, two weeks, until the partition is removed, until an upstream asset is removed(?), ...). One complication, I guess, is the interaction with AutoMaterialization policies though ...

**Reply by clairelin135** [reply]
> Do you know a reliable way to find the files associated with assets that have no partition anymore?

Do you mean the output files etc? Depending on your IO manager, those output files have been persisted somewhere. For example, with the default `fs_io_manager`, you can see the output files per-partition in the DAGSTER_HOME directory, even if those partitions no longer exist.

If you mean the runs/events, you can query the instance via `instance.get_run_records(RunsFilter(tags={"dagster/partition": partition_key}))` or `instance.get_event_records(EventRecordsFilter(asset_partitions=[...]))` to get the runs/events of prior partitions. Not 100% sure this would work, but in the runs page in the UI, you could also filter by a `dagster/partition` tag, which would allow you to see runs for partitions that are no longer part of your partitions def.

**Reply by cleboo** [reply]
I was able to retrieve the asset paths from the event log:
```python
with DagsterInstance.get() as instance:
    asset_partitions = set([
        (
            item.event_log_entry.dagster_event.event_specific_data.materialization.partition,
            item.event_log_entry.dagster_event.event_specific_data.materialization.metadata['path'].path,
        )
        for item in instance.get_event_records(EventRecordsFilter(
            DagsterEventType.ASSET_MATERIALIZATION,
            asset_key=AssetKey("my_asset")
        ))
    ])
# {('2023-07-16', '/path/to/my_asset/2023-07-16'),
#  ('2023-07-17', '/path/to/my_asset/2023-07-17'),
#  ('2023-07-18', '/path/to/my_asset/2023-07-18'),
# ...
```
The partitions can be fetched for comparison like so:
```python
with DagsterInstance.get() as instance:
    partitions = instance.get_dynamic_partitions("my_partition")
```
So I could remove the files and also the events from the log to wipe these assets. However, this only works with `fs_io_manager` and can not really be generalized because I could not find an indication of the `io_manager` that was used to save an asset. I think this should really be solved by the `io_manager` as indicated also here: https://github.com/dagster-io/dagster/discussions/14978

I submitted a feature request that outlines my rough idea for a solution to the "how long do we keep data?" problem: https://github.com/dagster-io/dagster/issues/15526

Thanks for helping me out!
